### ISSUE TITLE
BUG: TreeExplainer.shap_values() mutates expected_value for XGBoost models

### ISSUE THREAD

**Issue by BaibhavKundu2005** [shap/shap#4495]
### Issue Description

`TreeExplainer.expected_value` is computed incorrectly at initialization for certain XGBoost model types, and is silently corrected as a side effect of calling `shap_values()`. This means:
1. `expected_value` before calling `shap_values()` is wrong
2. `expected_value` after calling `shap_values()` is correct
3. Users who read `expected_value` before calling `shap_values()` get an incorrect value with no warning
4. The shape also changes from (1,) to () which is a secondary type inconsistency issue

### Minimal Reproducible Example

```python
import shap
import numpy as np
import xgboost as xgb
from sklearn.datasets import make_classification

X, y = make_classification(n_samples=200, n_features=10, random_state=42)
X_train, X_test = X[:150], X[150:]
y_train = y[:150]

model = xgb.XGBClassifier(n_estimators=50, random_state=42)
model.fit(X_train, y_train)

explainer = shap.TreeExplainer(model)

print(f"expected_value before shap_values() : {explainer.expected_value}")
print(f"shape before                         : {np.array(explainer.expected_value).shape}")

shap_values = explainer.shap_values(X_test)

print(f"expected_value after shap_values()   : {explainer.expected_value}")
print(f"shape after                          : {np.array(explainer.expected_value).shape}")

# confirm the post-call value is the correct one
dmatrix = xgb.DMatrix(X_test)
raw_margin = model.get_booster().predict(dmatrix, output_margin=True)
print(f"\nMean raw margin (correct baseline)   : {raw_margin.mean():.8f}")
print(f"expected_value before call           : {np.array(explainer.expected_value - shap_values.mean(0).sum()):.8f}")
```

### Traceback

```shell
No exception or warning is raised. The bug manifests silently as a value and shape change:

    expected_value before shap_values() : [-0.02666822]   shape=(1,)
    expected_value after  shap_values() : -0.01702357     shape=()

Users who read expected_value before calling shap_values() receive an incorrect value with no indication anything is wrong. Users who read it after receive the correct value. The shape inconsistency (1,) vs () is also present regardless of which value is correct.

Additionally confirmed in a real-world dataset by @khteh in #4414:

    Before shap_values(): [0.16934096]   shape=(1,)
    After  shap_values(): 0.18604861     shape=()

```

### Expected Behavior

`expected_value` should be correct immediately after `shap.TreeExplainer(model)` initialization, without requiring a `shap_values()` call first. Its value and shape should remain consistent regardless of whether `shap_values()` has been called.

This issue was first reported in #652 (June 2019) by @BasilBeirouti, who confirmed that for certain XGBoost models the initialization value is incorrect and only becomes correct after calling `shap_values()`. @slundberg tagged it as `todo` at the time:

> "This is something that needs investigated/potentially fixed, 
> so I am adding a todo tag for now."

It was observed again by @mueslo in October 2025 and independently rediscovered through investigation of #4414 in April 2026.

**Important note for anyone attempting a fix:** A naive guard of `if self.expected_value is None` at the mutation sites in 
`shap_values()` breaks additivity for multiple XGBoost model types (XGBRFRegressor, XGBClassifier, XGBRFClassifier, XGBRanker, tweedie regression, dart regression). This was confirmed by running the existing test suite after applying the guard, 10 XGBoost tests failed with additivity errors. The correct fix is to repair the initialization path, so it computes the correct value upfront, rather than preventing the correction during `shap_values()`.

The secondary shape inconsistency - `(1,)` before vs `()` after - should also be normalized as part of any fix, regardless of which value is correct.

### Bug report checklist

- [ ] I have checked that this issue has not already been reported.
- [x] I have confirmed this bug exists on the [latest release](https://github.com/shap/shap/releases) of shap.
- [x] I have confirmed this bug exists on the [master branch](https://github.com/shap/shap/blob/master/CONTRIBUTING.md#installing-from-the-master-branch) of shap.
- [x] I'd be interested in making a PR to fix this bug

### Installed Versions

`0.51.0`

**Comment by DashratRajpurohit** [comment]
I've submitted a PR to fix this: #4499 

**The approach:** Rather than preventing the silent correction inside
`shap_values()`, the fix repairs the initialization path so `expected_value`
is computed correctly upfront using XGBoost's own C++ `pred_contribs` output
— the same source that `shap_values()` has always used. This eliminates the
mutation entirely, which is the cleanest solution.

A safe fallback to the old manual calculation is kept for any edge cases.
I'm happy to iterate on this if the maintainers have a preferred approach.
